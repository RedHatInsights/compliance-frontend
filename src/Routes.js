import { Route, Switch, Redirect } from 'react-router-dom';
import React from 'react';
import asyncComponent from './Utilities/asyncComponent';
import some from 'lodash/some';

/**
 * Aysnc imports of components
 *
 * https://webpack.js.org/guides/code-splitting/
 * https://reactjs.org/docs/code-splitting.html
 *
 * pros:
 *      1) code splitting
 *      2) can be used in server-side rendering
 * cons:
 *      1) nameing chunk names adds unnecessary docs to code,
 *         see the difference with DashboardMap and InventoryDeployments.
 *
 */
const CompliancePolicies = asyncComponent(() =>
    import(/* webpackChunkName: "CompliancePolicies" */ 'SmartComponents/CompliancePolicies/CompliancePolicies')
);
const Reports = asyncComponent(() =>
    import(/* webpackChunkName: "ComplianceReports" */ './SmartComponents/Reports/Reports')
);
const ReportsSystems = asyncComponent(() =>
    import(/* webpackChunkName: "ComplianceReportsSystems" */ './SmartComponents/ReportsSystems/ReportsSystems')
);
const ReportDetails = asyncComponent(() =>
    import(/* webpackChunkName: "ComplianceReportDetails" */ './SmartComponents/ReportDetails/ReportDetails')
);
const ComplianceSystems = asyncComponent(() =>
    import(/* webpackChunkName: "ComplianceSystems" */ 'SmartComponents/ComplianceSystems/ComplianceSystems')
);
const PolicyDetails = asyncComponent(() =>
    import(/* webpackChunkName: "PolicyDetails" */ 'SmartComponents/PolicyDetails/PolicyDetails')
);
const SystemDetails = asyncComponent(() =>
    import(/* webpackChunkName: "SystemDetails" */ 'SmartComponents/SystemDetails/SystemDetails')
);

export const paths = {
    compliancePolicies: '/policies',
    reports: '/reports',
    reportsSystems: '/reports/systems',
    complianceSystems: '/systems',
    complianceSystemsInventoryDetail: '/systems/:inventoryId',
    policyDetails: '/policies/:policy_id',
    policyDetailsInventoryDetail: '/policies/:policy_id/:inventoryId',
    reportDetails: '/reports/:report_id',
    reportDetailsInventoryDetail: '/policies/:policy_id/:inventoryId',
    systemDetails: '/systems/:inventoryId'
};

type Props = {
    childProps: any
};

/**
 * the Switch component changes routes depending on the path.
 *
 * Route properties:
 *      exact - path must match exactly,
 *      path - https://prod.foo.redhat.com:1337/insights/compliance/rules
 *      component - component to be rendered when a route has been chosen.
 */
export const Routes = (props: Props) => {
    const path = props.childProps.location.pathname;

    return (
        <Switch>
            <Route exact path={paths.compliancePolicies} component={CompliancePolicies} />
            <Route exact path={paths.reports} component={Reports} />
            <Route exact path={paths.reportsSystems} component={ReportsSystems} />
            <Route exact path={paths.reportDetails} component={ReportDetails} />
            <Route exact path={paths.complianceSystems} component={ComplianceSystems} />
            <Route path={paths.complianceSystemsInventoryDetail} component={SystemDetails} />
            <Route path={paths.policyDetailsInventoryDetail} component={SystemDetails} />
            <Route exact path={paths.policyDetails} component={PolicyDetails} />
            <Route path={paths.systemDetails} component={SystemDetails} />

            {/* Finally, catch all unmatched routes */}
            <Route render={() => (some(paths, p => p === path) ? null : <Redirect to={paths.reports} />)} />
        </Switch>
    );
};
