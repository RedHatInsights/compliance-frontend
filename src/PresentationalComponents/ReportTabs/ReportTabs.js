import routerParams from '@redhat-cloud-services/frontend-components-utilities/files/RouterParams';
import propTypes from 'prop-types';
import React from 'react';
import { paths } from '../../Routes';
import { Tabs, Tab } from '@patternfly/react-core';
import invert from 'lodash/invert';

export const ReportTabs = (props) => {
    const { match: { path } } = props;

    const tabPaths = {
        0: paths.reports,
        1: paths.reportsSystems
    };

    const redirect = (_, tabIndex) => {
        props.history.push(tabPaths[tabIndex]);
    };

    let currentKey = Number(invert(tabPaths)[path]);

    const tabs = [
        <Tab title={'By policy'} key={0} eventKey={0}></Tab>,
        <Tab title={'By systems'} key={1} eventKey={1}></Tab>
    ];

    return (
        <Tabs activeKey={currentKey} onSelect={redirect} aria-label="Report Tabs">
            { tabs }
        </Tabs>
    );
};

ReportTabs.propTypes = {
    history: propTypes.object,
    match: propTypes.object
};

export default routerParams(ReportTabs);
