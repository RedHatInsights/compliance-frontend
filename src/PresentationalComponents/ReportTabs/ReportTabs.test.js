import toJson from 'enzyme-to-json';
import { paths } from '../../Routes';

import { ReportTabs } from './ReportTabs';

describe('ReportTabs', () => {
    const defaultProps = {
        history: {},
        match: {
            path: paths.complianceSystems
        }
    };

    it('expect to render without error', () => {
        const wrapper = shallow(
            <ReportTabs { ...defaultProps } />
        );

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    // TODO: Add test triggering a click on a tab to test "redirect"
});
