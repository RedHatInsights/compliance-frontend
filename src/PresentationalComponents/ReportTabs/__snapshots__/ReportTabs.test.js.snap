// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReportTabs expect to render without error 1`] = `
<Component
  activeKey={NaN}
  aria-label="Report Tabs"
  onSelect={[Function]}
>
  <Tab
    eventKey={0}
    key="0"
    title="By policy"
  />
  <Tab
    eventKey={1}
    key="1"
    title="By systems"
  />
</Component>
`;
